{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "64b5fc3b-807a-42da-b66d-d0826b7cc87c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "data = pd.ExcelFile(\"ARTIvir_CoV_minimal_combined_dset.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b58672b8-3411-4b73-b567-4c3d23164b4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "data.sheet_names;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8ee0a2de-5985-4e4d-ad96-862fdd5e3245",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs =  {sheet_name: data.parse(sheet_name) for sheet_name in data.sheet_names}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "357ae984-b259-4d11-ab57-56e41ca5a013",
   "metadata": {},
   "outputs": [],
   "source": [
    "virus_host_PPI = dfs[\"PPIs\"]\n",
    "effectome = dfs[\"EFF\"]\n",
    "proteome = dfs[\"NHBE_FPs\"]\n",
    "phosphoproteome = dfs[\"PPx\"]\n",
    "ubiquitinome = dfs[\"Ux\"]\n",
    "transcriptome = dfs[\"Tx\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc5becf3-7d18-4cd3-8d6a-c0cb5df5cd7d",
   "metadata": {},
   "source": [
    "# Proteome complete data analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b35371dc-fa64-4f32-b8bd-4f0a4c9be89f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Majority.protein.IDs</th>\n",
       "      <th>Gene.names</th>\n",
       "      <th>lambda.min.converged</th>\n",
       "      <th>Lasso_SARSCoV.min</th>\n",
       "      <th>Lasso_SARSCoV.min.p</th>\n",
       "      <th>Lasso_SARSCoV2.min</th>\n",
       "      <th>Lasso_SARSCoV2.min.p</th>\n",
       "      <th>lambda.min</th>\n",
       "      <th>external_synonym</th>\n",
       "      <th>Lasso_SARSCoV.min.sig</th>\n",
       "      <th>Lasso_SARSCoV2.min.sig</th>\n",
       "      <th>rn_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sp|A0A0U1RRL7|MMPOS_HUMAN</td>\n",
       "      <td>MMP24OS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.015477</td>\n",
       "      <td>MMP24OS;MMP24-AS1</td>\n",
       "      <td>not_sig</td>\n",
       "      <td>not_sig</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sp|A0AV96-2|RBM47_HUMAN;sp|A0AV96|RBM47_HUMAN</td>\n",
       "      <td>RBM47</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.614245</td>\n",
       "      <td>0.003332</td>\n",
       "      <td>0.913143</td>\n",
       "      <td>6.281934e-15</td>\n",
       "      <td>0.002456</td>\n",
       "      <td>RBM47;FLJ20273;NET18</td>\n",
       "      <td>low_sig</td>\n",
       "      <td>high_sig</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sp|A0AVT1|UBA6_HUMAN;sp|A0AVT1-2|UBA6_HUMAN</td>\n",
       "      <td>UBA6</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-0.042801</td>\n",
       "      <td>0.474667</td>\n",
       "      <td>-0.077215</td>\n",
       "      <td>1.452895e-01</td>\n",
       "      <td>0.000078</td>\n",
       "      <td>UBA6;FLJ10808;UBE1L2</td>\n",
       "      <td>not_sig</td>\n",
       "      <td>not_sig</td>\n",
       "      <td>UBA6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sp|A0FGR8-2|ESYT2_HUMAN;sp|A0FGR8|ESYT2_HUMAN;...</td>\n",
       "      <td>ESYT2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.015747</td>\n",
       "      <td>ESYT2;CHR2SYT;FAM62B;KIAA1228</td>\n",
       "      <td>not_sig</td>\n",
       "      <td>not_sig</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sp|A0JNW5|UH1BL_HUMAN</td>\n",
       "      <td>UHRF1BP1L</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.026649</td>\n",
       "      <td>UHRF1BP1L;KIAA0701</td>\n",
       "      <td>not_sig</td>\n",
       "      <td>not_sig</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                Majority.protein.IDs Gene.names  \\\n",
       "0                          sp|A0A0U1RRL7|MMPOS_HUMAN    MMP24OS   \n",
       "1      sp|A0AV96-2|RBM47_HUMAN;sp|A0AV96|RBM47_HUMAN      RBM47   \n",
       "2        sp|A0AVT1|UBA6_HUMAN;sp|A0AVT1-2|UBA6_HUMAN       UBA6   \n",
       "3  sp|A0FGR8-2|ESYT2_HUMAN;sp|A0FGR8|ESYT2_HUMAN;...      ESYT2   \n",
       "4                              sp|A0JNW5|UH1BL_HUMAN  UHRF1BP1L   \n",
       "\n",
       "   lambda.min.converged  Lasso_SARSCoV.min  Lasso_SARSCoV.min.p  \\\n",
       "0                   1.0           0.000000                  NaN   \n",
       "1                   1.0           0.614245             0.003332   \n",
       "2                   1.0          -0.042801             0.474667   \n",
       "3                   1.0           0.000000                  NaN   \n",
       "4                   1.0           0.000000                  NaN   \n",
       "\n",
       "   Lasso_SARSCoV2.min  Lasso_SARSCoV2.min.p  lambda.min  \\\n",
       "0            0.000000                   NaN    0.015477   \n",
       "1            0.913143          6.281934e-15    0.002456   \n",
       "2           -0.077215          1.452895e-01    0.000078   \n",
       "3            0.000000                   NaN    0.015747   \n",
       "4            0.000000                   NaN    0.026649   \n",
       "\n",
       "                external_synonym Lasso_SARSCoV.min.sig Lasso_SARSCoV2.min.sig  \\\n",
       "0              MMP24OS;MMP24-AS1               not_sig                not_sig   \n",
       "1           RBM47;FLJ20273;NET18               low_sig               high_sig   \n",
       "2           UBA6;FLJ10808;UBE1L2               not_sig                not_sig   \n",
       "3  ESYT2;CHR2SYT;FAM62B;KIAA1228               not_sig                not_sig   \n",
       "4             UHRF1BP1L;KIAA0701               not_sig                not_sig   \n",
       "\n",
       "  rn_name  \n",
       "0     NaN  \n",
       "1     NaN  \n",
       "2    UBA6  \n",
       "3     NaN  \n",
       "4     NaN  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proteome.head()\n",
    "#len(proteome[\"Gene.names\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b08717a9-1d8e-4fbd-b15e-6ae149d8ecc5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "complete data for proteome in SARS-CoV 3166 on a total of 6129\n"
     ]
    }
   ],
   "source": [
    "sars_cov_proteome_nan = proteome[\"Lasso_SARSCoV.min.p\"].isnull()\n",
    "print(f\"complete data for proteome in SARS-CoV {len(sars_cov_proteome_nan) - sars_cov_proteome_nan.sum()} on a total of {len(sars_cov_proteome_nan)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "37286cf5-e749-4ed1-bc95-3ab0446f9521",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "complete data for proteome in SARS-CoV 3812 on a total of 6129\n"
     ]
    }
   ],
   "source": [
    "sars_cov_2_proteome_nan = proteome[\"Lasso_SARSCoV2.min.p\"].isnull()\n",
    "print(f\"complete data for proteome in SARS-CoV {len(sars_cov_2_proteome_nan) - sars_cov_2_proteome_nan.sum()} on a total of {len(sars_cov_2_proteome_nan)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8dd60538-b202-4005-aef1-79b4b1b1b244",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3166"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sars_cov_complete_genes = proteome[\"Gene.names\"][sars_cov_proteome_nan == False]\n",
    "len(sars_cov_complete_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a3b8cd76-f3fc-4c5c-a9e9-16c17622265f",
   "metadata": {},
   "outputs": [],
   "source": [
    "sars_cov_2_complete_genes = proteome[\"Gene.names\"][sars_cov_2_proteome_nan == False]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "97277615-4c7a-49b6-92e9-5226050ebe9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_proteins_list = proteome[\"Gene.names\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b389d9b-0e62-4165-b517-57c8dfe3f759",
   "metadata": {},
   "source": [
    "# Transcriptome complete data analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ceb6d693-ee37-4ea4-9c31-660a72889e8c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "13108"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcriptome.head()\n",
    "len(transcriptome[\"gene_name\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2caccab0-deec-40c5-af56-ecca6a143377",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of nan values for SARS-CoV-2 0\n",
      "number of nan values for SARS-CoV 0\n"
     ]
    }
   ],
   "source": [
    "num_nan_values_sars_cov_2 = transcriptome[\"fold_change_log2.SARS_CoV2@6h_vs_mock@6h\"].isnull().sum() + transcriptome[\"fold_change_log2.SARS_CoV2@12h_vs_mock@12h\"].isnull().sum() + transcriptome[\"fold_change_log2.SARS_CoV2@24h_vs_mock@24h\"].isnull().sum()\n",
    "num_nan_values_sars_cov = transcriptome[\"fold_change_log2.SARS_CoV@6h_vs_mock@6h\"].isnull().sum() + transcriptome[\"fold_change_log2.SARS_CoV@12h_vs_mock@12h\"].isnull().sum() + transcriptome[\"fold_change_log2.SARS_CoV@24h_vs_mock@24h\"].isnull().sum()\n",
    "\n",
    "print(f\"number of nan values for SARS-CoV-2 {num_nan_values_sars_cov_2}\")\n",
    "print(f\"number of nan values for SARS-CoV {num_nan_values_sars_cov}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7196a20f-f65f-4b26-a663-dce099d912a6",
   "metadata": {},
   "source": [
    "# P-value definition:\n",
    "## the probability that a random sample from the posterior distribution of the first condition would be smaller (or larger) than a random sample drawn from the second condition. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99c7288e-3d76-4328-9717-9a0b12f795cf",
   "metadata": {},
   "source": [
    "Case of SARS-CoV-2:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "3c622ae3-fd03-4f3a-9e85-f04c8a7e24ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "In SARS-CoV-2 we have regulated genes with p-value < 0.05 shared among the time points 6h,12h,24h = 200 on a total of 1249 at 6h of 1778 at 12h of 1399 at 24h\n",
      "Unique genes with good p-values 3475 on a total of 4426\n"
     ]
    }
   ],
   "source": [
    "def omic_completeness_analyzer(omic, which_sars: str):\n",
    "    gene_list_6h = omic[\"gene_name\"][omic[\"p_value.\" + which_sars + \"@6h_vs_mock@6h\"] < 0.05].values\n",
    "    gene_list_12h = omic[\"gene_name\"][omic[\"p_value.\" + which_sars + \"@12h_vs_mock@12h\"] < 0.05].values\n",
    "    gene_list_24h = omic[\"gene_name\"][omic[\"p_value.\" + which_sars + \"@24h_vs_mock@24h\"] < 0.05].values\n",
    "\n",
    "    import numpy as np\n",
    "    shared_genes_with_good_p_value = []\n",
    "    for item in gene_list_6h:\n",
    "        if (item in gene_list_12h) & (item in gene_list_24h):\n",
    "            shared_genes_with_good_p_value.append(item)\n",
    "    shared_genes_with_good_p_value = np.array(shared_genes_with_good_p_value)\n",
    "    \n",
    "    unique_genes_with_good_p_value = []\n",
    "    for item in gene_list_6h:\n",
    "        if item not in unique_genes_with_good_p_value:\n",
    "            unique_genes_with_good_p_value.append(item)\n",
    "    for item in gene_list_12h:\n",
    "        if item not in unique_genes_with_good_p_value:\n",
    "            unique_genes_with_good_p_value.append(item) \n",
    "    for item in gene_list_24h:\n",
    "        if item not in unique_genes_with_good_p_value:\n",
    "            unique_genes_with_good_p_value.append(item)        \n",
    "    return gene_list_6h, gene_list_12h, gene_list_24h, shared_genes_with_good_p_value, unique_genes_with_good_p_value\n",
    "\n",
    "gene_list_6h_cov2, gene_list_12h_cov2, gene_list_24h_cov2, shared_genes_with_good_p_value_cov2, unique_genes_with_good_p_value_cov2  = omic_completeness_analyzer(transcriptome, \"SARS_CoV2\")\n",
    "print(f\"In SARS-CoV-2 we have regulated genes with p-value < 0.05 shared among the time points 6h,12h,24h = {len(shared_genes_with_good_p_value_cov2)} on a total of {len(gene_list_6h_cov2)} at 6h of {len(gene_list_12h_cov2)} at 12h of {len(gene_list_24h_cov2)} at 24h\")\n",
    "#transcriptome[transcriptome[\"gene_name\"] == \"AL022718.1\"][[\"p_value.SARS_CoV2@6h_vs_mock@6h\", \"fold_change_log2.SARS_CoV2@6h_vs_mock@6h\", \"p_value.SARS_CoV2@12h_vs_mock@12h\", \"fold_change_log2.SARS_CoV2@12h_vs_mock@12h\", \"p_value.SARS_CoV2@24h_vs_mock@24h\", \"fold_change_log2.SARS_CoV2@24h_vs_mock@24h\"]]\n",
    "print(f\"Unique genes with good p-values {len(unique_genes_with_good_p_value_cov2)} on a total of {len(gene_list_6h_cov2) + len(gene_list_12h_cov2) + len(gene_list_24h_cov2)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c014591-6e56-424f-a26f-0a6b5b46b35b",
   "metadata": {},
   "source": [
    "Case of SARS-CoV:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "9b262dee-5ad0-4435-882d-6b14b7c107ae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "In SARS-CoV we have regulated genes with p-value < 0.05 shared among the time points 6h,12h,24h = 212 on a total of 1059 at 6h of 1597 at 12h of 2327 at 24h\n",
      "Unique genes with good p-values 3876 on a total of 4983\n"
     ]
    }
   ],
   "source": [
    "gene_list_6h_cov, gene_list_12h_cov, gene_list_24h_cov, shared_genes_with_good_p_value_cov, unique_genes_with_good_p_value_cov = omic_completeness_analyzer(transcriptome, \"SARS_CoV\")\n",
    "print(f\"In SARS-CoV we have regulated genes with p-value < 0.05 shared among the time points 6h,12h,24h = {len(shared_genes_with_good_p_value_cov)} on a total of {len(gene_list_6h_cov)} at 6h of {len(gene_list_12h_cov)} at 12h of {len(gene_list_24h_cov)} at 24h\")\n",
    "print(f\"Unique genes with good p-values {len(unique_genes_with_good_p_value_cov)} on a total of {len(gene_list_6h_cov) + len(gene_list_12h_cov) + len(gene_list_24h_cov)}\")\n",
    "\n",
    "#transcriptome[transcriptome[\"gene_name\"] == \"AL022718.1\"][[\"p_value.SARS_CoV2@6h_vs_mock@6h\", \"fold_change_log2.SARS_CoV2@6h_vs_mock@6h\", \"p_value.SARS_CoV2@12h_vs_mock@12h\", \"fold_change_log2.SARS_CoV2@12h_vs_mock@12h\", \"p_value.SARS_CoV2@24h_vs_mock@24h\", \"fold_change_log2.SARS_CoV2@24h_vs_mock@24h\"]]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "597c11cf-dbdc-4edc-b2c5-6ffb656cb31b",
   "metadata": {},
   "source": [
    "# Ubiquitinome complete data analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "9b319c35-5817-48c2-b2bb-1527afbe5074",
   "metadata": {},
   "outputs": [],
   "source": [
    "ubiquitinome.columns;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "44e292b5-b120-42ea-abf5-14280393a19d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of nan values for SARS-CoV-2 0\n",
      "number of nan values for SARS-CoV 0\n"
     ]
    }
   ],
   "source": [
    "num_nan_values_sars_cov_2 = ubiquitinome[\"fold_change_log2.SARS_CoV2@6h_vs_mock@6h\"].isnull().sum() + ubiquitinome[\"fold_change_log2.SARS_CoV2@12h_vs_mock@12h\"].isnull().sum() + ubiquitinome[\"fold_change_log2.SARS_CoV2@24h_vs_mock@24h\"].isnull().sum()\n",
    "num_nan_values_sars_cov = ubiquitinome[\"fold_change_log2.SARS_CoV@6h_vs_mock@6h\"].isnull().sum() + ubiquitinome[\"fold_change_log2.SARS_CoV@12h_vs_mock@12h\"].isnull().sum() + ubiquitinome[\"fold_change_log2.SARS_CoV@24h_vs_mock@24h\"].isnull().sum()\n",
    "\n",
    "print(f\"number of nan values for SARS-CoV-2 {num_nan_values_sars_cov_2}\")\n",
    "print(f\"number of nan values for SARS-CoV {num_nan_values_sars_cov}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ed96fc7-f01b-4597-95ef-38dee129b6e2",
   "metadata": {},
   "source": [
    "Case of SARS-CoV-2:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "c23ca858-fb2c-41c7-8e0f-b3801212589e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "In SARS-CoV-2 we have ubiquitinome regulated genes with p-value < 0.05 shared among the time points 6h,12h,24h = 166 on a total of 240 at 6h of 791 at 12h of 3181 at 24h\n",
      "Unique genes with good p-values 1610 on a total of 4212\n"
     ]
    }
   ],
   "source": [
    "gene_list_6h_cov2, gene_list_12h_cov2, gene_list_24h_cov2, ubi_shared_genes_with_good_p_value_cov2, ubi_unique_genes_with_good_p_value_cov2 = omic_completeness_analyzer(ubiquitinome, \"SARS_CoV2\")\n",
    "print(f\"In SARS-CoV-2 we have ubiquitinome regulated genes with p-value < 0.05 shared among the time points 6h,12h,24h = {len(ubi_shared_genes_with_good_p_value_cov2)} on a total of {len(gene_list_6h_cov2)} at 6h of {len(gene_list_12h_cov2)} at 12h of {len(gene_list_24h_cov2)} at 24h\")\n",
    "print(f\"Unique genes with good p-values {len(ubi_unique_genes_with_good_p_value_cov2)} on a total of {len(gene_list_6h_cov2) + len(gene_list_12h_cov2) + len(gene_list_24h_cov2)}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4d13239-8c6a-4a56-9ccd-b5861f33a775",
   "metadata": {},
   "source": [
    "Case of SARS-CoV:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "3509283b-0a3a-40f9-9ccb-0abb1dcb3ef1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "In SARS-CoV we have ubiquitinome regulated genes with p-value < 0.05 shared among the time points 6h,12h,24h = 181 on a total of 290 at 6h of 669 at 12h of 2259 at 24h\n",
      "Unique genes with good p-values 1315 on a total of 3218\n"
     ]
    }
   ],
   "source": [
    "gene_list_6h_cov, gene_list_12h_cov, gene_list_24h_cov, ubi_shared_genes_with_good_p_value_cov, ubi_unique_genes_with_good_p_value_cov = omic_completeness_analyzer(ubiquitinome, \"SARS_CoV\")\n",
    "print(f\"In SARS-CoV we have ubiquitinome regulated genes with p-value < 0.05 shared among the time points 6h,12h,24h = {len(ubi_shared_genes_with_good_p_value_cov)} on a total of {len(gene_list_6h_cov)} at 6h of {len(gene_list_12h_cov)} at 12h of {len(gene_list_24h_cov)} at 24h\")\n",
    "print(f\"Unique genes with good p-values {len(ubi_unique_genes_with_good_p_value_cov)} on a total of {len(gene_list_6h_cov) + len(gene_list_12h_cov) + len(gene_list_24h_cov)}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52d202e0-54f8-4897-83f0-e5862b9b30f7",
   "metadata": {},
   "source": [
    "# Phosphoproteome complete data analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "5ac80019-8422-4f08-9015-8f6e36e3b1fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "phosphoproteome.columns;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "25fe73d0-9d06-45f3-87cf-809496220f9b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of nan values for SARS-CoV-2 0\n",
      "number of nan values for SARS-CoV 0\n"
     ]
    }
   ],
   "source": [
    "num_nan_values_sars_cov_2 = phosphoproteome[\"fold_change_log2.SARS_CoV2@6h_vs_mock@6h\"].isnull().sum() + phosphoproteome[\"fold_change_log2.SARS_CoV2@12h_vs_mock@12h\"].isnull().sum() + phosphoproteome[\"fold_change_log2.SARS_CoV2@24h_vs_mock@24h\"].isnull().sum()\n",
    "num_nan_values_sars_cov = phosphoproteome[\"fold_change_log2.SARS_CoV@6h_vs_mock@6h\"].isnull().sum() + phosphoproteome[\"fold_change_log2.SARS_CoV@12h_vs_mock@12h\"].isnull().sum() + phosphoproteome[\"fold_change_log2.SARS_CoV@24h_vs_mock@24h\"].isnull().sum()\n",
    "\n",
    "print(f\"number of nan values for SARS-CoV-2 {num_nan_values_sars_cov_2}\")\n",
    "print(f\"number of nan values for SARS-CoV {num_nan_values_sars_cov}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eece25ad-7bf4-42a0-9db7-b4ff6bee2cf1",
   "metadata": {},
   "source": [
    "Case of SARS-CoV-2:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "93fd63a7-75d9-4fc4-92b7-02372ad2f5e6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "In SARS-CoV-2 we have phosphoproteome regulated genes with p-value < 0.05 shared among the time points 6h,12h,24h = 1491 on a total of 1743 at 6h of 3284 at 12h of 5645 at 24h\n",
      "Unique genes with good p-values 2604 on a total of 10672\n"
     ]
    }
   ],
   "source": [
    "gene_list_6h_cov2, gene_list_12h_cov2, gene_list_24h_cov2, pho_shared_genes_with_good_p_value_cov2, pho_unique_genes_with_good_p_value_cov2 = omic_completeness_analyzer(phosphoproteome, \"SARS_CoV2\")\n",
    "print(f\"In SARS-CoV-2 we have phosphoproteome regulated genes with p-value < 0.05 shared among the time points 6h,12h,24h = {len(pho_shared_genes_with_good_p_value_cov2)} on a total of {len(gene_list_6h_cov2)} at 6h of {len(gene_list_12h_cov2)} at 12h of {len(gene_list_24h_cov2)} at 24h\")\n",
    "print(f\"Unique genes with good p-values {len(pho_unique_genes_with_good_p_value_cov2)} on a total of {len(gene_list_6h_cov2) + len(gene_list_12h_cov2) + len(gene_list_24h_cov2)}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9746a94-7042-43bd-8660-b514fcd73d13",
   "metadata": {},
   "source": [
    "Case of SARS-CoV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "f3c6a973-a58b-4a85-a583-eceac88e9249",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "In SARS-CoV we have phosphoproteome regulated genes with p-value < 0.05 shared among the time points 6h,12h,24h = 1074 on a total of 1316 at 6h of 2593 at 12h of 4726 at 24h\n",
      "Unique genes with good p-values 2326 on a total of 8635\n"
     ]
    }
   ],
   "source": [
    "gene_list_6h_cov, gene_list_12h_cov, gene_list_24h_cov, pho_shared_genes_with_good_p_value_cov, pho_unique_genes_with_good_p_value_cov = omic_completeness_analyzer(phosphoproteome, \"SARS_CoV\")\n",
    "print(f\"In SARS-CoV we have phosphoproteome regulated genes with p-value < 0.05 shared among the time points 6h,12h,24h = {len(pho_shared_genes_with_good_p_value_cov)} on a total of {len(gene_list_6h_cov)} at 6h of {len(gene_list_12h_cov)} at 12h of {len(gene_list_24h_cov)} at 24h\")\n",
    "print(f\"Unique genes with good p-values {len(pho_unique_genes_with_good_p_value_cov)} on a total of {len(gene_list_6h_cov) + len(gene_list_12h_cov) + len(gene_list_24h_cov)}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba1fe4d6-a75d-4574-a521-2878109a8dd5",
   "metadata": {},
   "source": [
    "## Dataset creation SARS-CoV-2\n",
    "\n",
    "effectome, proteome, phosphoproteome, ubiquitinome, transcriptome. \n",
    "Lets collect all the genes found in either one of the omics."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "40d44199-eb67-4181-badd-25bb08ffb1c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "def add_string_items_only(list_to_fill, list_to_check):\n",
    "    for item in list_to_check:\n",
    "        if type(item) == str:\n",
    "            list_to_fill.append(item)\n",
    "\n",
    "effectome_host_genes = []\n",
    "add_string_items_only(effectome_host_genes, effectome[effectome[\"Viral protein\"] == \"SARS-CoV-2\"][\"Host protein\"].unique()[1:])\n",
    "        \n",
    "proteome_host_genes = []\n",
    "add_string_items_only(proteome_host_genes, proteome[\"Gene.names\"].unique())\n",
    "                     \n",
    "phosphoproteome_host_genes = []\n",
    "add_string_items_only(phosphoproteome_host_genes, phosphoproteome[phosphoproteome[\"is_viral\"] == False][\"gene_name\"].unique())\n",
    "                     \n",
    "ubiquitinome_host_genes = []\n",
    "add_string_items_only(ubiquitinome_host_genes, ubiquitinome[ubiquitinome[\"is_viral\"] == False][\"gene_name\"].unique())\n",
    "        \n",
    "transcriptome_host_genes = []\n",
    "add_string_items_only(transcriptome_host_genes, transcriptome[\"gene_name\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0ec3913e-7623-44a4-847c-0d5fd6c0aa5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_protein_coding_genes = np.unique(np.append(np.append(np.append(effectome_host_genes, proteome_host_genes), phosphoproteome_host_genes),ubiquitinome_host_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b9b1a667-3af2-4a04-92e2-86c139e23e98",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "eccomi\n"
     ]
    }
   ],
   "source": [
    "for item in all_protein_coding_genes:\n",
    "    if item == \"RPS17\":\n",
    "        print(\"eccomi\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a46b2649-b9a0-48b0-8714-38909b1183ab",
   "metadata": {},
   "source": [
    "Let's start creating the most complete dataset, \n",
    "- setting all the missing feature data to zero\n",
    "- deleting all the proteins found in STRING PPI network but not in my most complete list of proteins taken from all the omics  \n",
    "\n",
    "I start by downloading the database of the human STRING protein IDs with the corresponding gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a7698f7b-607a-4e9e-9fef-398c9bed59c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pybiomart import Server, Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2378404d-9d39-4f3e-8165-c04e38cdbaeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = Dataset(name='hsapiens_gene_ensembl',\n",
    "                  host='http://www.ensembl.org')\n",
    "\n",
    "server = Server(host='http://www.ensembl.org')\n",
    "dataset_server =  (server.marts['ENSEMBL_MART_ENSEMBL'].datasets['hsapiens_gene_ensembl'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e8fc09a8-84a1-44d9-b1cb-edfd89b57580",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_from_protein_id_to_gene_name = dataset_server.query(attributes=['ensembl_peptide_id', 'external_gene_name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d9eef840-b12e-43cd-b453-ea5ca5d72050",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>MT-TF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>MT-RNR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>MT-TV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>MT-RNR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>MT-TL1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Protein stable ID Gene name\n",
       "0               NaN     MT-TF\n",
       "1               NaN   MT-RNR1\n",
       "2               NaN     MT-TV\n",
       "3               NaN   MT-RNR2\n",
       "4               NaN    MT-TL1"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_from_protein_id_to_gene_name.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "51c00c15-9f13-4bdf-8cfd-522210d70394",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_from_protein_id_to_gene_name = dataset.query(attributes=['ensembl_peptide_id', 'external_gene_name'])\n",
    "#df_from_protein_id_to_gene_name.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58b903f6-3b11-4513-8afb-7afaa82fd73b",
   "metadata": {},
   "source": [
    "- I need now to import the PPI network from STRING"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3d9b52b0-a71d-4b3a-8b1d-ea2cd47f629e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000264718</td>\n",
       "      <td>157</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000346046</td>\n",
       "      <td>178</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000347134</td>\n",
       "      <td>143</td>\n",
       "      <td>0</td>\n",
       "      <td>63</td>\n",
       "      <td>162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000317469</td>\n",
       "      <td>345</td>\n",
       "      <td>0</td>\n",
       "      <td>92</td>\n",
       "      <td>379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000302393</td>\n",
       "      <td>288</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>287</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               protein1              protein2  experimental  database  \\\n",
       "0  9606.ENSP00000000233  9606.ENSP00000264718           157         0   \n",
       "1  9606.ENSP00000000233  9606.ENSP00000346046           178         0   \n",
       "2  9606.ENSP00000000233  9606.ENSP00000347134           143         0   \n",
       "3  9606.ENSP00000000233  9606.ENSP00000317469           345         0   \n",
       "4  9606.ENSP00000000233  9606.ENSP00000302393           288         0   \n",
       "\n",
       "   textmining  combined_score  \n",
       "0           0             156  \n",
       "1           0             177  \n",
       "2          63             162  \n",
       "3          92             379  \n",
       "4           0             287  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_string_ppi = pd.read_csv(\"9606.protein.physical.links.detailed.v11.5.txt\", sep = \" \")\n",
    "df_string_ppi.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2718fa0-64e1-4efe-82c1-4c512fc63bb8",
   "metadata": {},
   "source": [
    "- remove all the lines with $ experimental < 10$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4afa3ff7-c1a6-4e2d-8330-878d8d8cab6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1991832"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_string_ppi[\"experimental\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "171c8c06-bc86-4a09-85ca-b6299c32405f",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10000\n",
      "20000\n",
      "30000\n",
      "40000\n",
      "50000\n",
      "60000\n",
      "70000\n",
      "80000\n",
      "90000\n",
      "100000\n",
      "110000\n",
      "120000\n",
      "130000\n",
      "140000\n",
      "150000\n",
      "160000\n",
      "170000\n",
      "180000\n",
      "190000\n",
      "200000\n",
      "210000\n",
      "220000\n",
      "230000\n",
      "236632\n"
     ]
    }
   ],
   "source": [
    "counter = 0\n",
    "index_list = []\n",
    "for index, row in df_string_ppi.iterrows():\n",
    "    if row[\"experimental\"] < 10:\n",
    "        counter += 1\n",
    "        index_list.append(index)\n",
    "        if counter % 10000 == 0:    \n",
    "            print(counter)\n",
    "        \n",
    "print(counter)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3e627c4a-4eb0-4bb7-afc2-376406cf1213",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_string_ppi.drop(index_list, axis = 0, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "89ca834c-e78b-412a-920f-0dcc29fb495e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_string_ppi.reset_index(drop = True, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "86aba696-3bfa-424d-a2fd-4f38e5185aec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000264718</td>\n",
       "      <td>157</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000346046</td>\n",
       "      <td>178</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000347134</td>\n",
       "      <td>143</td>\n",
       "      <td>0</td>\n",
       "      <td>63</td>\n",
       "      <td>162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000317469</td>\n",
       "      <td>345</td>\n",
       "      <td>0</td>\n",
       "      <td>92</td>\n",
       "      <td>379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000302393</td>\n",
       "      <td>288</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>287</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               protein1              protein2  experimental  database  \\\n",
       "0  9606.ENSP00000000233  9606.ENSP00000264718           157         0   \n",
       "1  9606.ENSP00000000233  9606.ENSP00000346046           178         0   \n",
       "2  9606.ENSP00000000233  9606.ENSP00000347134           143         0   \n",
       "3  9606.ENSP00000000233  9606.ENSP00000317469           345         0   \n",
       "4  9606.ENSP00000000233  9606.ENSP00000302393           288         0   \n",
       "\n",
       "   textmining  combined_score  \n",
       "0           0             156  \n",
       "1           0             177  \n",
       "2          63             162  \n",
       "3          92             379  \n",
       "4           0             287  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_string_ppi.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "fb11662f-524d-4f97-9886-b3af212e21a9",
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "index 0 is out of bounds for axis 0 with size 0",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-14-23273fd6cbc4>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdf_from_protein_id_to_gene_name\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf_from_protein_id_to_gene_name\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"Protein stable ID\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"ENSP00000346046\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"Gene name\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m: index 0 is out of bounds for axis 0 with size 0"
     ]
    }
   ],
   "source": [
    "df_from_protein_id_to_gene_name[df_from_protein_id_to_gene_name[\"Protein stable ID\"] == \"ENSP00000346046\"][\"Gene name\"].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "fc49167f-b1c8-4cac-b16b-e3e08c146af6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "147208    ENSP00000248450\n",
       "147209    ENSP00000403343\n",
       "147210                NaN\n",
       "147211    ENSP00000416394\n",
       "147212    ENSP00000396295\n",
       "147213    ENSP00000393818\n",
       "Name: Protein stable ID, dtype: object"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_from_protein_id_to_gene_name[df_from_protein_id_to_gene_name[\"Gene name\"] == \"AAMP\"][\"Protein stable ID\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89b26354-5ab5-4d61-b0db-b8dafd4d9cfe",
   "metadata": {},
   "source": [
    "- ensamble.org is not complete enough, proteins present in STRING PPI network and in my list of \"all_protein_coding_genes\" are not listed in ensamble\n",
    "- I will create a dictionary protein IDs to Gene Names from uniprot database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f6b7c70c-552d-4c02-ad88-a7f4bb9b4304",
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio.SwissProt as sp\n",
    "import csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "27dff055-330b-430f-bcf2-871297022ec4",
   "metadata": {},
   "outputs": [],
   "source": [
    "datContent = [i.strip().split() for i in open(\"./HUMAN_9606_idmapping.dat\").readlines()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "49418ccc-603c-4810-99e5-0cea477c4a62",
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter gene names and STRING names\n",
    "string_and_gene_names = []\n",
    "for item in datContent:\n",
    "    if (item[1] == \"Gene_Name\") or (item[1] == \"STRING\"):\n",
    "        string_and_gene_names.append(item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7964afe1-7562-44a9-90bf-1dcc3bcb1947",
   "metadata": {},
   "outputs": [],
   "source": [
    "uniprot_names = []\n",
    "for item in string_and_gene_names:\n",
    "    uniprot_names.append(item[0])\n",
    "    \n",
    "used_names = set()    \n",
    "unique_uniprot_names = [elem for elem in uniprot_names if elem not in used_names and (used_names.add(elem) or True)]    \n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d770e546-75d4-4410-aaf8-a13bf423b613",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10000\n",
      "20000\n",
      "30000\n",
      "40000\n",
      "50000\n",
      "60000\n",
      "70000\n",
      "80000\n",
      "90000\n",
      "100000\n",
      "110000\n",
      "120000\n",
      "130000\n",
      "140000\n",
      "150000\n",
      "160000\n",
      "170000\n",
      "180000\n"
     ]
    }
   ],
   "source": [
    "string_names_list = []\n",
    "gene_names_list = []\n",
    "counter = 0\n",
    "for elem in string_and_gene_names:\n",
    "    if elem[1] == \"STRING\":\n",
    "        for names in string_and_gene_names:\n",
    "            if names[0] == elem[0] and names[1] == \"Gene_Name\":\n",
    "                string_names_list.append(elem[2])\n",
    "                gene_names_list.append(names[2])\n",
    "                break\n",
    "    counter += 1\n",
    "    if counter % 10000 == 0:\n",
    "        print(counter)\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "bdd96256-3589-44cd-bb32-7576ef42c0c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "string_to_gen_names_map = {\"string_ids\" : string_names_list, \"gene_names\" : gene_names_list}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "3b0635be-3030-4e1c-9522-41cec94c2408",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_string_to_gen_names_map = pd.DataFrame(string_to_gen_names_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "7e228d14-42b6-4c6a-b1b1-5dde99a44322",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>string_ids</th>\n",
       "      <th>gene_names</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606.ENSP00000361930</td>\n",
       "      <td>YWHAB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606.ENSP00000264335</td>\n",
       "      <td>YWHAE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606.ENSP00000248975</td>\n",
       "      <td>YWHAH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606.ENSP00000306330</td>\n",
       "      <td>YWHAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606.ENSP00000340989</td>\n",
       "      <td>SFN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             string_ids gene_names\n",
       "0  9606.ENSP00000361930      YWHAB\n",
       "1  9606.ENSP00000264335      YWHAE\n",
       "2  9606.ENSP00000248975      YWHAH\n",
       "3  9606.ENSP00000306330      YWHAG\n",
       "4  9606.ENSP00000340989        SFN"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_string_to_gen_names_map.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "ef41837c-709a-4724-b18e-a2c358f3c467",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000264718</td>\n",
       "      <td>157</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000346046</td>\n",
       "      <td>178</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000347134</td>\n",
       "      <td>143</td>\n",
       "      <td>0</td>\n",
       "      <td>63</td>\n",
       "      <td>162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000317469</td>\n",
       "      <td>345</td>\n",
       "      <td>0</td>\n",
       "      <td>92</td>\n",
       "      <td>379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000302393</td>\n",
       "      <td>288</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>287</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               protein1              protein2  experimental  database  \\\n",
       "0  9606.ENSP00000000233  9606.ENSP00000264718           157         0   \n",
       "1  9606.ENSP00000000233  9606.ENSP00000346046           178         0   \n",
       "2  9606.ENSP00000000233  9606.ENSP00000347134           143         0   \n",
       "3  9606.ENSP00000000233  9606.ENSP00000317469           345         0   \n",
       "4  9606.ENSP00000000233  9606.ENSP00000302393           288         0   \n",
       "\n",
       "   textmining  combined_score  \n",
       "0           0             156  \n",
       "1           0             177  \n",
       "2          63             162  \n",
       "3          92             379  \n",
       "4           0             287  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_string_ppi.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "2fd0b85d-c4a3-4c2a-aa6a-9089f9b1df33",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'RPS17'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_string_to_gen_names_map[df_string_to_gen_names_map[\"string_ids\"] == \"9606.ENSP00000346046\"][\"gene_names\"].values[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8675129b-8742-4fb2-84a9-a24b7f6302db",
   "metadata": {},
   "source": [
    "- let's create the gene names list corresponding to columns \"protein1\" and \"protein2\" in df_string_ppi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "4907e6c6-d753-48b5-bb9e-53f9f95e35e9",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "10000\n",
      "20000\n",
      "30000\n",
      "40000\n",
      "50000\n",
      "60000\n",
      "70000\n",
      "80000\n",
      "90000\n",
      "100000\n",
      "110000\n",
      "120000\n",
      "130000\n",
      "140000\n",
      "150000\n",
      "160000\n",
      "170000\n",
      "180000\n",
      "190000\n",
      "200000\n",
      "210000\n",
      "220000\n",
      "230000\n",
      "240000\n",
      "250000\n",
      "260000\n",
      "270000\n",
      "280000\n",
      "290000\n",
      "300000\n",
      "310000\n",
      "320000\n",
      "330000\n",
      "340000\n",
      "350000\n",
      "360000\n",
      "370000\n",
      "380000\n",
      "390000\n",
      "400000\n",
      "410000\n",
      "420000\n",
      "430000\n",
      "440000\n",
      "450000\n",
      "460000\n",
      "470000\n",
      "480000\n",
      "490000\n",
      "500000\n",
      "510000\n",
      "520000\n",
      "530000\n",
      "540000\n",
      "550000\n",
      "560000\n",
      "570000\n",
      "580000\n",
      "590000\n",
      "600000\n",
      "610000\n",
      "620000\n",
      "630000\n",
      "640000\n",
      "650000\n",
      "660000\n",
      "670000\n",
      "680000\n",
      "690000\n",
      "700000\n",
      "710000\n",
      "720000\n",
      "730000\n",
      "740000\n",
      "750000\n",
      "760000\n",
      "770000\n",
      "780000\n",
      "790000\n",
      "800000\n",
      "810000\n",
      "820000\n",
      "830000\n",
      "840000\n",
      "850000\n",
      "860000\n",
      "870000\n",
      "880000\n",
      "890000\n",
      "900000\n",
      "910000\n",
      "920000\n",
      "930000\n",
      "940000\n",
      "950000\n",
      "960000\n",
      "970000\n",
      "980000\n",
      "990000\n",
      "1000000\n",
      "1010000\n",
      "1020000\n",
      "1030000\n",
      "1040000\n",
      "1050000\n",
      "1060000\n",
      "1070000\n",
      "1080000\n",
      "1090000\n",
      "1100000\n",
      "1110000\n",
      "1120000\n",
      "1130000\n",
      "1140000\n",
      "1150000\n",
      "1160000\n",
      "1170000\n",
      "1180000\n",
      "1190000\n",
      "1200000\n",
      "1210000\n",
      "1220000\n",
      "1230000\n",
      "1240000\n",
      "1250000\n",
      "1260000\n",
      "1270000\n",
      "1280000\n",
      "1290000\n",
      "1300000\n",
      "1310000\n",
      "1320000\n",
      "1330000\n",
      "1340000\n",
      "1350000\n",
      "1360000\n",
      "1370000\n",
      "1380000\n",
      "1390000\n",
      "1400000\n",
      "1410000\n",
      "1420000\n",
      "1430000\n",
      "1440000\n",
      "1450000\n",
      "1460000\n",
      "1470000\n",
      "1480000\n",
      "1490000\n",
      "1500000\n",
      "1510000\n",
      "1520000\n",
      "1530000\n",
      "1540000\n",
      "1550000\n",
      "1560000\n",
      "1570000\n",
      "1580000\n",
      "1590000\n",
      "1600000\n",
      "1610000\n",
      "1620000\n",
      "1630000\n",
      "1640000\n",
      "1650000\n",
      "1660000\n",
      "1670000\n",
      "1680000\n",
      "1690000\n",
      "1700000\n",
      "1710000\n",
      "1720000\n",
      "1730000\n",
      "1740000\n",
      "1750000\n"
     ]
    }
   ],
   "source": [
    "gene_names_1 = []\n",
    "gene_names_2 = []\n",
    "\n",
    "for index, row in df_string_ppi.iterrows():\n",
    "    names_1 = df_string_to_gen_names_map[df_string_to_gen_names_map[\"string_ids\"] == row[\"protein1\"]][\"gene_names\"]\n",
    "    names_2 = df_string_to_gen_names_map[df_string_to_gen_names_map[\"string_ids\"] == row[\"protein2\"]][\"gene_names\"]\n",
    "    if len(names_1) > 0:\n",
    "        gene_names_1.append(names_1.values[0])\n",
    "    else:\n",
    "        gene_names_1.append('nan')\n",
    "    if len(names_2) > 0:    \n",
    "        gene_names_2.append(names_2.values[0])\n",
    "    else:\n",
    "        gene_names_2.append('nan')\n",
    "    if index % 10000 == 0:\n",
    "        print(index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "c6461916-f78c-4a08-817d-10e16e3d9a53",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Percentage of not itentified gene names in STRING PPI dataset: 2.682543299908842%\n"
     ]
    }
   ],
   "source": [
    "counter = 0\n",
    "for item in gene_names_2:\n",
    "    if item == \"nan\":\n",
    "        counter += 1\n",
    "print(f\"Percentage of not itentified gene names in STRING PPI dataset: {100*counter/len(gene_names_2)}%\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "67c72f5e-a53b-454e-9c98-1177e62e98d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_string_ppi[\"gene_name_1\"] = pd.Series(gene_names_1, index = df_string_ppi.index)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efe8ba75-e8ee-4439-8bca-2714817302d9",
   "metadata": {},
   "source": [
    "- I add the corresponding gene names to the protein interaction network "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "b62d7876-09c0-48ba-8930-44308fdf2917",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>gene_name_1</th>\n",
       "      <th>gene_name_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000264718</td>\n",
       "      <td>157</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>156</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>GPN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000346046</td>\n",
       "      <td>178</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>177</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>RPS17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000347134</td>\n",
       "      <td>143</td>\n",
       "      <td>0</td>\n",
       "      <td>63</td>\n",
       "      <td>162</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>NET1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000317469</td>\n",
       "      <td>345</td>\n",
       "      <td>0</td>\n",
       "      <td>92</td>\n",
       "      <td>379</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>BBS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000302393</td>\n",
       "      <td>288</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>287</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>LDHAL6B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               protein1              protein2  experimental  database  \\\n",
       "0  9606.ENSP00000000233  9606.ENSP00000264718           157         0   \n",
       "1  9606.ENSP00000000233  9606.ENSP00000346046           178         0   \n",
       "2  9606.ENSP00000000233  9606.ENSP00000347134           143         0   \n",
       "3  9606.ENSP00000000233  9606.ENSP00000317469           345         0   \n",
       "4  9606.ENSP00000000233  9606.ENSP00000302393           288         0   \n",
       "\n",
       "   textmining  combined_score gene_name_1 gene_name_2  \n",
       "0           0             156        ARF5        GPN1  \n",
       "1           0             177        ARF5       RPS17  \n",
       "2          63             162        ARF5        NET1  \n",
       "3          92             379        ARF5        BBS1  \n",
       "4           0             287        ARF5     LDHAL6B  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_string_ppi[\"gene_name_2\"] = pd.Series(gene_names_2, index = df_string_ppi.index)\n",
    "df_string_ppi.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "97881548-b86d-40bf-be05-2038309f67ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the protein interaction network to a file \n",
    "#df_string_ppi.to_csv('df_string_ppi.zip', index=False, compression = dict(method='zip',archive_name='df_string_ppi.csv'))\n",
    "df_string_ppi = pd.read_csv(\"df_string_ppi.zip\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "016234cd-9681-41ef-bcfa-84bc428f8697",
   "metadata": {},
   "outputs": [],
   "source": [
    "transcriptome_host_genes = np.array(transcriptome_host_genes[:-2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "07b3d0eb-fd0f-475a-8f73-3e3bad6ee08a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9467"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_genes = np.unique(np.append(all_protein_coding_genes, all_protein_coding_genes))\n",
    "len(all_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53611e60-c62f-4656-a513-e8af4c28fdd0",
   "metadata": {},
   "source": [
    "- check gene names in PPI network that are not present in my list of gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "62403094-ca12-444a-86dc-285b40267f5d",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/samuele/.pyenv/versions/3.6.10/envs/jupiter-env/lib/python3.6/site-packages/ipykernel_launcher.py:4: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison\n",
      "  after removing the cwd from sys.path.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1000\n",
      "2000\n",
      "3000\n",
      "4000\n",
      "5000\n",
      "6000\n",
      "7000\n",
      "8000\n",
      "9000\n",
      "10000\n",
      "11000\n",
      "12000\n",
      "13000\n",
      "14000\n",
      "15000\n",
      "16000\n",
      "17000\n",
      "18000\n",
      "19000\n",
      "20000\n",
      "21000\n",
      "22000\n",
      "23000\n",
      "24000\n",
      "25000\n",
      "26000\n",
      "27000\n",
      "28000\n",
      "29000\n",
      "30000\n",
      "31000\n",
      "32000\n",
      "33000\n",
      "34000\n",
      "35000\n",
      "36000\n",
      "37000\n",
      "38000\n",
      "39000\n",
      "40000\n",
      "41000\n",
      "42000\n",
      "43000\n",
      "44000\n",
      "45000\n",
      "46000\n",
      "47000\n",
      "48000\n",
      "49000\n",
      "50000\n",
      "51000\n",
      "52000\n",
      "53000\n",
      "54000\n",
      "55000\n",
      "56000\n",
      "57000\n",
      "58000\n",
      "59000\n",
      "60000\n",
      "61000\n",
      "62000\n",
      "63000\n",
      "64000\n",
      "65000\n",
      "66000\n",
      "67000\n",
      "68000\n",
      "69000\n",
      "70000\n",
      "71000\n",
      "72000\n",
      "73000\n",
      "74000\n",
      "75000\n",
      "76000\n",
      "77000\n",
      "78000\n",
      "79000\n",
      "80000\n",
      "81000\n",
      "82000\n",
      "83000\n",
      "84000\n",
      "85000\n",
      "86000\n",
      "87000\n",
      "88000\n",
      "89000\n",
      "90000\n",
      "91000\n",
      "92000\n",
      "93000\n",
      "94000\n",
      "95000\n",
      "96000\n",
      "97000\n",
      "98000\n",
      "99000\n",
      "100000\n",
      "101000\n",
      "102000\n",
      "103000\n",
      "104000\n",
      "105000\n",
      "106000\n",
      "107000\n",
      "108000\n",
      "109000\n",
      "110000\n",
      "111000\n",
      "112000\n",
      "113000\n",
      "114000\n",
      "115000\n",
      "116000\n",
      "117000\n",
      "118000\n",
      "119000\n",
      "120000\n",
      "121000\n",
      "122000\n",
      "123000\n",
      "124000\n",
      "125000\n",
      "126000\n",
      "127000\n",
      "128000\n",
      "129000\n",
      "130000\n",
      "131000\n",
      "132000\n",
      "133000\n",
      "134000\n",
      "135000\n",
      "136000\n",
      "137000\n",
      "138000\n",
      "139000\n",
      "140000\n",
      "141000\n",
      "142000\n",
      "143000\n",
      "144000\n",
      "145000\n",
      "146000\n",
      "147000\n",
      "148000\n",
      "149000\n",
      "150000\n",
      "151000\n",
      "152000\n",
      "153000\n",
      "154000\n",
      "155000\n",
      "156000\n",
      "157000\n",
      "158000\n",
      "159000\n",
      "160000\n",
      "161000\n",
      "162000\n",
      "163000\n",
      "164000\n",
      "165000\n",
      "166000\n",
      "167000\n",
      "168000\n",
      "169000\n",
      "170000\n",
      "171000\n",
      "172000\n",
      "173000\n",
      "174000\n",
      "175000\n",
      "176000\n",
      "177000\n",
      "178000\n",
      "179000\n",
      "180000\n",
      "181000\n",
      "182000\n",
      "183000\n",
      "184000\n",
      "185000\n",
      "186000\n",
      "187000\n",
      "188000\n",
      "189000\n",
      "190000\n",
      "191000\n",
      "192000\n",
      "193000\n",
      "194000\n",
      "195000\n",
      "196000\n",
      "197000\n",
      "198000\n",
      "199000\n",
      "200000\n",
      "201000\n",
      "202000\n",
      "203000\n",
      "204000\n",
      "205000\n",
      "206000\n",
      "207000\n",
      "208000\n",
      "209000\n",
      "210000\n",
      "211000\n",
      "212000\n",
      "213000\n",
      "214000\n",
      "215000\n",
      "216000\n",
      "217000\n",
      "218000\n",
      "219000\n",
      "220000\n",
      "221000\n",
      "222000\n",
      "223000\n",
      "224000\n",
      "225000\n",
      "226000\n",
      "227000\n",
      "228000\n",
      "229000\n",
      "230000\n",
      "231000\n",
      "232000\n",
      "233000\n",
      "234000\n",
      "235000\n",
      "236000\n",
      "237000\n",
      "238000\n",
      "239000\n",
      "240000\n",
      "241000\n",
      "242000\n",
      "243000\n",
      "244000\n",
      "245000\n",
      "246000\n",
      "247000\n",
      "248000\n",
      "249000\n",
      "250000\n",
      "251000\n",
      "252000\n",
      "253000\n",
      "254000\n",
      "255000\n",
      "256000\n",
      "257000\n",
      "258000\n",
      "259000\n",
      "260000\n",
      "261000\n",
      "262000\n",
      "263000\n",
      "264000\n",
      "265000\n",
      "266000\n",
      "267000\n",
      "268000\n",
      "269000\n",
      "270000\n",
      "271000\n",
      "272000\n",
      "273000\n",
      "274000\n",
      "275000\n",
      "276000\n",
      "277000\n",
      "278000\n",
      "279000\n",
      "280000\n",
      "281000\n",
      "282000\n",
      "283000\n",
      "284000\n",
      "285000\n",
      "286000\n",
      "287000\n",
      "288000\n",
      "289000\n",
      "290000\n",
      "291000\n",
      "292000\n",
      "293000\n",
      "294000\n",
      "295000\n",
      "296000\n",
      "297000\n",
      "298000\n",
      "299000\n",
      "300000\n",
      "301000\n",
      "302000\n",
      "303000\n",
      "304000\n",
      "305000\n",
      "306000\n",
      "307000\n",
      "308000\n",
      "309000\n",
      "310000\n",
      "311000\n",
      "312000\n",
      "313000\n",
      "314000\n",
      "315000\n",
      "316000\n",
      "317000\n",
      "318000\n",
      "319000\n",
      "320000\n",
      "321000\n",
      "322000\n",
      "323000\n",
      "324000\n",
      "325000\n",
      "326000\n",
      "327000\n",
      "328000\n",
      "329000\n",
      "330000\n",
      "331000\n",
      "332000\n",
      "333000\n",
      "334000\n",
      "335000\n",
      "336000\n",
      "337000\n",
      "338000\n",
      "339000\n",
      "340000\n",
      "341000\n",
      "342000\n",
      "343000\n",
      "344000\n",
      "345000\n",
      "346000\n",
      "347000\n",
      "348000\n",
      "349000\n",
      "350000\n",
      "351000\n",
      "352000\n",
      "353000\n",
      "354000\n",
      "355000\n",
      "356000\n",
      "357000\n",
      "358000\n",
      "359000\n",
      "360000\n",
      "361000\n",
      "362000\n",
      "363000\n",
      "364000\n",
      "365000\n",
      "366000\n",
      "367000\n",
      "368000\n",
      "369000\n",
      "370000\n",
      "371000\n",
      "372000\n",
      "373000\n",
      "374000\n",
      "375000\n",
      "376000\n",
      "377000\n",
      "378000\n",
      "379000\n",
      "380000\n",
      "381000\n",
      "382000\n",
      "383000\n",
      "384000\n",
      "385000\n",
      "386000\n",
      "387000\n",
      "388000\n",
      "389000\n",
      "390000\n",
      "391000\n",
      "392000\n",
      "393000\n",
      "394000\n",
      "395000\n",
      "396000\n",
      "397000\n",
      "398000\n",
      "399000\n",
      "400000\n",
      "401000\n",
      "402000\n",
      "403000\n",
      "404000\n",
      "405000\n",
      "406000\n",
      "407000\n",
      "408000\n",
      "409000\n",
      "410000\n",
      "411000\n",
      "412000\n",
      "413000\n",
      "414000\n",
      "415000\n",
      "416000\n",
      "417000\n",
      "418000\n",
      "419000\n",
      "420000\n",
      "421000\n",
      "422000\n",
      "423000\n",
      "424000\n",
      "425000\n",
      "426000\n",
      "427000\n",
      "428000\n",
      "429000\n",
      "430000\n",
      "431000\n",
      "432000\n",
      "433000\n",
      "434000\n",
      "435000\n",
      "436000\n",
      "437000\n",
      "438000\n",
      "439000\n",
      "440000\n",
      "441000\n",
      "442000\n",
      "443000\n",
      "444000\n",
      "445000\n",
      "446000\n",
      "447000\n",
      "448000\n",
      "449000\n",
      "450000\n",
      "451000\n",
      "452000\n",
      "453000\n",
      "454000\n",
      "455000\n",
      "456000\n",
      "457000\n",
      "458000\n",
      "459000\n",
      "460000\n",
      "461000\n",
      "462000\n",
      "463000\n",
      "464000\n",
      "465000\n",
      "466000\n",
      "467000\n",
      "468000\n",
      "469000\n",
      "470000\n",
      "471000\n",
      "472000\n",
      "473000\n",
      "474000\n",
      "475000\n",
      "476000\n",
      "477000\n",
      "478000\n",
      "479000\n",
      "480000\n",
      "481000\n",
      "482000\n",
      "483000\n",
      "484000\n",
      "485000\n",
      "486000\n",
      "487000\n",
      "488000\n",
      "489000\n",
      "490000\n",
      "491000\n",
      "492000\n",
      "493000\n",
      "494000\n",
      "495000\n",
      "496000\n",
      "497000\n",
      "498000\n",
      "499000\n",
      "500000\n",
      "501000\n",
      "502000\n",
      "503000\n",
      "504000\n",
      "505000\n",
      "506000\n",
      "507000\n",
      "508000\n",
      "509000\n",
      "510000\n",
      "511000\n",
      "512000\n",
      "513000\n",
      "514000\n",
      "515000\n",
      "516000\n",
      "517000\n",
      "518000\n",
      "519000\n",
      "520000\n",
      "521000\n",
      "522000\n",
      "523000\n",
      "524000\n",
      "525000\n",
      "526000\n",
      "527000\n",
      "528000\n",
      "529000\n",
      "530000\n",
      "531000\n",
      "532000\n",
      "533000\n",
      "534000\n",
      "535000\n",
      "536000\n",
      "537000\n",
      "538000\n",
      "539000\n",
      "540000\n",
      "541000\n",
      "542000\n",
      "543000\n",
      "544000\n",
      "545000\n",
      "546000\n",
      "547000\n",
      "548000\n",
      "549000\n",
      "550000\n",
      "551000\n",
      "552000\n",
      "553000\n",
      "554000\n",
      "555000\n",
      "556000\n",
      "557000\n",
      "558000\n",
      "559000\n",
      "560000\n",
      "561000\n",
      "562000\n",
      "563000\n",
      "564000\n",
      "565000\n",
      "566000\n",
      "567000\n",
      "568000\n",
      "569000\n",
      "570000\n",
      "571000\n",
      "572000\n",
      "573000\n",
      "574000\n",
      "575000\n",
      "576000\n",
      "577000\n",
      "578000\n",
      "579000\n",
      "580000\n",
      "581000\n",
      "582000\n",
      "583000\n",
      "584000\n",
      "585000\n",
      "586000\n",
      "587000\n",
      "588000\n",
      "589000\n",
      "590000\n",
      "591000\n",
      "592000\n",
      "593000\n",
      "594000\n",
      "595000\n",
      "596000\n",
      "597000\n",
      "598000\n",
      "599000\n",
      "600000\n",
      "601000\n",
      "602000\n",
      "603000\n",
      "604000\n",
      "605000\n",
      "606000\n",
      "607000\n",
      "608000\n",
      "609000\n",
      "610000\n",
      "611000\n",
      "612000\n",
      "613000\n",
      "614000\n",
      "615000\n",
      "616000\n",
      "617000\n",
      "618000\n",
      "619000\n",
      "620000\n",
      "621000\n",
      "622000\n",
      "623000\n",
      "624000\n",
      "625000\n",
      "626000\n",
      "627000\n",
      "628000\n",
      "629000\n",
      "630000\n",
      "631000\n",
      "632000\n",
      "633000\n",
      "634000\n",
      "635000\n",
      "636000\n",
      "637000\n",
      "638000\n",
      "639000\n",
      "640000\n",
      "641000\n",
      "642000\n",
      "643000\n",
      "644000\n",
      "645000\n",
      "646000\n",
      "647000\n",
      "648000\n",
      "649000\n",
      "650000\n",
      "651000\n",
      "652000\n",
      "653000\n",
      "654000\n",
      "655000\n",
      "656000\n",
      "657000\n",
      "658000\n",
      "659000\n",
      "660000\n",
      "661000\n",
      "662000\n",
      "663000\n",
      "664000\n",
      "665000\n",
      "666000\n",
      "667000\n",
      "668000\n",
      "669000\n",
      "670000\n",
      "671000\n",
      "672000\n",
      "673000\n",
      "674000\n",
      "675000\n",
      "676000\n",
      "677000\n",
      "678000\n",
      "679000\n",
      "680000\n",
      "681000\n",
      "682000\n",
      "683000\n",
      "684000\n",
      "685000\n",
      "686000\n",
      "687000\n",
      "688000\n",
      "689000\n",
      "690000\n",
      "691000\n",
      "692000\n",
      "693000\n",
      "694000\n",
      "695000\n",
      "696000\n",
      "697000\n",
      "698000\n",
      "699000\n",
      "700000\n",
      "701000\n",
      "702000\n",
      "703000\n",
      "704000\n",
      "705000\n",
      "706000\n",
      "707000\n",
      "708000\n",
      "709000\n",
      "710000\n",
      "711000\n",
      "712000\n",
      "713000\n",
      "714000\n",
      "715000\n",
      "716000\n",
      "717000\n",
      "718000\n",
      "719000\n",
      "720000\n",
      "721000\n",
      "722000\n",
      "723000\n",
      "724000\n",
      "725000\n",
      "726000\n",
      "727000\n",
      "728000\n",
      "729000\n",
      "730000\n",
      "731000\n",
      "732000\n",
      "733000\n",
      "734000\n",
      "735000\n",
      "736000\n",
      "737000\n",
      "738000\n",
      "739000\n",
      "740000\n",
      "741000\n",
      "742000\n",
      "743000\n",
      "744000\n",
      "745000\n",
      "746000\n",
      "747000\n",
      "748000\n",
      "749000\n",
      "750000\n",
      "751000\n",
      "752000\n",
      "753000\n",
      "754000\n",
      "755000\n",
      "756000\n",
      "757000\n",
      "758000\n",
      "759000\n",
      "760000\n",
      "761000\n",
      "762000\n",
      "763000\n",
      "764000\n",
      "765000\n",
      "766000\n",
      "767000\n",
      "768000\n",
      "769000\n",
      "770000\n",
      "771000\n",
      "772000\n",
      "773000\n",
      "774000\n",
      "775000\n",
      "776000\n",
      "777000\n",
      "778000\n",
      "779000\n",
      "780000\n",
      "781000\n",
      "782000\n",
      "783000\n",
      "784000\n",
      "785000\n",
      "786000\n",
      "787000\n",
      "788000\n",
      "789000\n",
      "790000\n",
      "791000\n",
      "792000\n",
      "793000\n",
      "794000\n",
      "795000\n",
      "796000\n",
      "797000\n",
      "798000\n",
      "799000\n",
      "800000\n",
      "801000\n",
      "802000\n",
      "803000\n",
      "804000\n",
      "805000\n",
      "806000\n",
      "807000\n",
      "808000\n",
      "809000\n",
      "810000\n",
      "811000\n",
      "812000\n",
      "813000\n",
      "814000\n",
      "815000\n",
      "816000\n",
      "817000\n",
      "818000\n",
      "819000\n",
      "820000\n",
      "821000\n",
      "822000\n",
      "823000\n",
      "824000\n",
      "825000\n",
      "826000\n",
      "827000\n",
      "828000\n",
      "829000\n",
      "830000\n",
      "831000\n",
      "832000\n",
      "833000\n",
      "834000\n",
      "835000\n",
      "836000\n",
      "837000\n",
      "838000\n",
      "839000\n",
      "840000\n",
      "841000\n",
      "842000\n",
      "843000\n",
      "844000\n",
      "845000\n",
      "846000\n",
      "847000\n",
      "848000\n",
      "849000\n",
      "850000\n",
      "850234\n"
     ]
    }
   ],
   "source": [
    "counter = 0\n",
    "list_of_row_indices_to_drop = []\n",
    "for index, row in df_string_ppi.iterrows():\n",
    "    if (row[\"gene_name_1\"] not in all_genes) or (row[\"gene_name_2\"] not in all_genes):\n",
    "        counter += 1\n",
    "        list_of_row_indices_to_drop.append(index)\n",
    "        if counter % 1000 == 0:    \n",
    "            print(counter)\n",
    "        \n",
    "print(counter)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "5fc92a65-d5fb-416b-adc0-bde3caa0e739",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Percentage of lines to drop from the PPI network: 48.440861440291705%\n"
     ]
    }
   ],
   "source": [
    "print(f\"Percentage of lines to drop from the PPI network: {100*len(list_of_row_indices_to_drop)/len(df_string_ppi)}%\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c7d7c54-8077-4efd-bd49-16b608a28492",
   "metadata": {},
   "source": [
    "- I drop the links with proteins not present at all in my dataset "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "de6d2b9c-2d5e-4d58-957c-9cbc54a6285d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_string_ppi.drop(list_of_row_indices_to_drop, axis = 0, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "2ea34f6b-13c2-41cb-ba33-bae186a9775a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_string_ppi.reset_index(drop = True, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "7865fcfd-9177-43d2-b978-1d73b5829ab5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>gene_name_1</th>\n",
       "      <th>gene_name_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000264718</td>\n",
       "      <td>157</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>156</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>GPN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000346046</td>\n",
       "      <td>178</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>177</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>RPS17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000347134</td>\n",
       "      <td>143</td>\n",
       "      <td>0</td>\n",
       "      <td>63</td>\n",
       "      <td>162</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>NET1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000354876</td>\n",
       "      <td>153</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>152</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>MT-CO2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000381607</td>\n",
       "      <td>169</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>169</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>GSTP1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               protein1              protein2  experimental  database  \\\n",
       "0  9606.ENSP00000000233  9606.ENSP00000264718           157         0   \n",
       "1  9606.ENSP00000000233  9606.ENSP00000346046           178         0   \n",
       "2  9606.ENSP00000000233  9606.ENSP00000347134           143         0   \n",
       "3  9606.ENSP00000000233  9606.ENSP00000354876           153         0   \n",
       "4  9606.ENSP00000000233  9606.ENSP00000381607           169         0   \n",
       "\n",
       "   textmining  combined_score gene_name_1 gene_name_2  \n",
       "0           0             156        ARF5        GPN1  \n",
       "1           0             177        ARF5       RPS17  \n",
       "2          63             162        ARF5        NET1  \n",
       "3           0             152        ARF5      MT-CO2  \n",
       "4           0             169        ARF5       GSTP1  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_string_ppi.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "90687e1a-88f6-4270-9794-19ae7814ba0a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of links 904966\n",
      "Total number of nodes 8957\n",
      "Total number of starting proteins for which I have at least some feature data 9467\n"
     ]
    }
   ],
   "source": [
    "print(f\"Total number of links {len(df_string_ppi)}\")\n",
    "node_genes = np.unique(np.append(df_string_ppi[\"gene_name_2\"].unique(), df_string_ppi[\"gene_name_1\"].unique()))\n",
    "print(f\"Total number of nodes {len(node_genes)}\")\n",
    "print(f\"Total number of starting proteins for which I have at least some feature data {len(all_genes)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7aecaee-0b2f-442b-b0ad-da743f1b24a9",
   "metadata": {},
   "source": [
    "- let's evaluate the percentage of not identified genes using \"ensamble.org\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "93a73041-e81a-461f-a8cd-91e390c3f92e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_string_ppi_original = pd.read_csv(\"df_string_ppi.zip\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "a5cecd13-2c7a-4c44-924a-7ccc36a4cb1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1755200"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_string_ppi_original)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "e3fa4a94-66bd-4d89-9790-3f39de390cca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>gene_name_1</th>\n",
       "      <th>gene_name_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000264718</td>\n",
       "      <td>157</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>156</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>GPN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000346046</td>\n",
       "      <td>178</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>177</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>RPS17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000347134</td>\n",
       "      <td>143</td>\n",
       "      <td>0</td>\n",
       "      <td>63</td>\n",
       "      <td>162</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>NET1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000317469</td>\n",
       "      <td>345</td>\n",
       "      <td>0</td>\n",
       "      <td>92</td>\n",
       "      <td>379</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>BBS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000302393</td>\n",
       "      <td>288</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>287</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>LDHAL6B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               protein1              protein2  experimental  database  \\\n",
       "0  9606.ENSP00000000233  9606.ENSP00000264718           157         0   \n",
       "1  9606.ENSP00000000233  9606.ENSP00000346046           178         0   \n",
       "2  9606.ENSP00000000233  9606.ENSP00000347134           143         0   \n",
       "3  9606.ENSP00000000233  9606.ENSP00000317469           345         0   \n",
       "4  9606.ENSP00000000233  9606.ENSP00000302393           288         0   \n",
       "\n",
       "   textmining  combined_score gene_name_1 gene_name_2  \n",
       "0           0             156        ARF5        GPN1  \n",
       "1           0             177        ARF5       RPS17  \n",
       "2          63             162        ARF5        NET1  \n",
       "3          92             379        ARF5        BBS1  \n",
       "4           0             287        ARF5     LDHAL6B  "
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_string_ppi_original.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b91df00c-cf39-4ee2-bd28-c6c51fc6181b",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "10000\n",
      "20000\n",
      "30000\n",
      "40000\n",
      "50000\n",
      "60000\n",
      "70000\n",
      "80000\n",
      "90000\n",
      "100000\n",
      "110000\n",
      "120000\n",
      "130000\n",
      "140000\n",
      "150000\n",
      "160000\n",
      "170000\n",
      "180000\n",
      "190000\n",
      "200000\n",
      "210000\n",
      "220000\n",
      "230000\n",
      "240000\n",
      "250000\n",
      "260000\n",
      "270000\n",
      "280000\n",
      "290000\n",
      "300000\n",
      "310000\n",
      "320000\n",
      "330000\n",
      "340000\n",
      "350000\n",
      "360000\n",
      "370000\n",
      "380000\n",
      "390000\n",
      "400000\n",
      "410000\n",
      "420000\n",
      "430000\n",
      "440000\n",
      "450000\n",
      "460000\n",
      "470000\n",
      "480000\n",
      "490000\n",
      "500000\n",
      "510000\n",
      "520000\n",
      "530000\n",
      "540000\n",
      "550000\n",
      "560000\n",
      "570000\n",
      "580000\n",
      "590000\n",
      "600000\n",
      "610000\n",
      "620000\n",
      "630000\n",
      "640000\n",
      "650000\n",
      "660000\n",
      "670000\n",
      "680000\n",
      "690000\n",
      "700000\n",
      "710000\n",
      "720000\n",
      "730000\n",
      "740000\n",
      "750000\n",
      "760000\n",
      "770000\n",
      "780000\n",
      "790000\n",
      "800000\n",
      "810000\n",
      "820000\n",
      "830000\n",
      "840000\n",
      "850000\n",
      "860000\n",
      "870000\n",
      "880000\n"
     ]
    }
   ],
   "source": [
    "gene_names_from_ensamble_1 = []\n",
    "gene_names_from_ensamble_2 = []\n",
    "\n",
    "for index, row in df_string_ppi.iterrows():\n",
    "    names_1 = df_from_protein_id_to_gene_name[df_from_protein_id_to_gene_name[\"Protein stable ID\"] == row[\"protein1\"]][\"Gene name\"]\n",
    "    names_2 = df_from_protein_id_to_gene_name[df_from_protein_id_to_gene_name[\"Protein stable ID\"] == row[\"protein2\"]][\"Gene name\"]\n",
    "    if len(names_1) > 0:\n",
    "        gene_names_from_ensamble_1.append(names_1.values[0])\n",
    "    else:\n",
    "        gene_names_from_ensamble_1.append('nan')\n",
    "    if len(names_2) > 0:    \n",
    "        gene_names_from_ensamble_2.append(names_2.values[0])\n",
    "    else:\n",
    "        gene_names_from_ensamble_2.append('nan')\n",
    "    if index % 10000 == 0:\n",
    "        print(index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1035a49e-8225-4b41-bd98-130526586b72",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
